### ISSUE TITLE
RFC: Add Support for Well-Known URL for Changing Passwords

### ISSUE THREAD

**Issue by phiw13** [textpattern/textpattern#1746]
See the discussion in this related forum [thread](https://forum.textpattern.com/viewtopic.php?id=51645)

> `/.well-know/change-password` that would redirect to `textpattern/?reset=1`

This is mostly for the benefit of password managers: to make it easier for the user to update/modify a password. See the screenshot attached below for the Safari 14.1+ password manager.

spec text: https://w3c.github.io/webappsec-change-password-url/

This has been supported by Safari for a _quite_ a while and Chrome ( I think, based on [this](https://www.chromestatus.com/feature/6256768407568384) ).

to which @Bloke suggested

> Adding to .htaccess is an option but that leaves other server technology out in the cold. Since we can’t rely on its presence in, say, Nginx or IIS, it means that links to /well-known will then fail for such users, leaving them unable to change their passwords without manual URL intervention. So I would suggest we bake it into core’s admin-side URL router.

![password-manager](https://user-images.githubusercontent.com/8944809/139176311-fdca35d6-84ed-4cc6-a3f1-24e23c79a087.jpg)

**Comment by petecooper** [comment]
To get the RFC ball rolling…
After consideration, this is a soft -1 for me. I can see the benefits of it from a UX angle, but not in core and _on_ by default – perhaps deployed in a web server config, or – at a push – somewhere tucked away in advanced prefs and _off_ by default.
Considering there are tools that can scan the entire IPv4 space in under an hour (e.g. https://zmap.io), I don't want opportunists finding out what I'm running behind the scenes. And yes, I realise it's trivial to find out what's running by other means…but I don't personally want another way via a well-known (literally) URL.

**Comment by Bloke** [comment]
I'm not sure either. Just because doing something makes it easier for people, doesn't necessarily mean it's a good idea. It's easier to unlock your phone with a thumbprint but it's a terrible idea from a security standpoint compared to using a password/code.

If we do this, I'm with Pete that documenting how to do it is better than adding it as a feature. But I could be swayed if there's a compelling reason to make it more easily configurable. We have the option of adding it as a toggle in config.php. Perhaps the presence of that could add it to the routing table in publish.php so the redirect takes place? Not really thought it through.

**Comment by phiw13** [comment]
The “privacy” argument of hiding what (software) you are running I consider still completely silly.

> thumbprint but it's a terrible idea from a security standpoint compared to using a password/code.

Is it? people then set 1234 as a password. I am _very_ happy that my MBA has a touchID sensor to unlock it.

But OK, at the very least, documentation + a toggle in config.php is one step forward.

**Comment by Bloke** [comment]
> people then set 1234 as a password

Then they deserve to be hacked! But at least they can change it to something better when they do, and learn from their mistakes.

The overarching best security practice for at least the last 20+ years has been "never use the same password for more than one account". Now we're encouraged to use the same "password" (thumbprint) to unlock multiple accounts: your phone, your computer, your house, your burglar alarm, your car, and goodness knows what else in future... and that single password can never be changed so once it's known, that's it.

Anyway, back on topic, I agree security through obscurity doesn't hold much weight. But that doesn't stop us striving to minimise or reduce what we advertise. You never know when some [vulnerability might be discovered](https://www.theguardian.com/technology/2021/dec/10/software-flaw-most-critical-vulnerability-log-4-shell) that has people scrambling to find systems that use it to exploit it. If we make things harder by making the stealthiest codebase we can by default, that feels inherently safer somehow. I'm still looking for the best ways to reduce or eradicate our various 'tells' as we iterate Textpattern. There are a handful that are hard to code around, but over time we can gradually design the majority out.

An opt-in mechanism for those people that want to expose such information seems like the sanest way, imo. Bonus points if we could find a way to enable the `/.well-known` path in our routing, but then leave the redirection up to the administrator. A completely open example:

 ``` php
$txpcfg['redirects'] = array(
   '.well-known/dnt-policy.txt' => '/policies/do-not-track',
   '.well-known/change-password' => '/textpattern?reset=1',
...
);
```

That would also allow you to set up Textpattern-level redirects to just about anything by editing your config.php. Not pretty, but flexible. And there's nothing to stop us outputting this in Diagnostics if we think it might be useful as a debugging aid when someone says "my site is doing weird redirects".

Like I say, not thought it through so it might be untenable.

**Comment by petecooper** [comment]
>The “privacy” argument of hiding what (software) you are running I consider still completely silly.

Privacy, security, something else - pick your terminology.

In isolation, sure, not something you should necessarily rely on, but when there are sites that list the technology behind various websites _whether the admins like it or not_, that's overreaching. Heck, Bert obtained a list of some sites running Textpattern from BuiltWith…if memory serves he implied it didn't even cost anything beyond a polite email request…if it's that easy, and a zero day CVE pops up with perhaps an RCE for a Textpattern site, it's trivial to connect a list of sites to a bot and -- short of a well-configured WAF -- game over for a lot of Textpattern sites. It's happened to Drupal (e.g. Drupalgeddon) and, sure, there are many more Drupal eyeballs + instances around the web, and it was a mess at the time.

An admin can choose to remove `meta` stuff from `head`, prevent access to the files that are unique-ish to Textpattern and give the version number away (I won't mention what they are), install a WAF, and be clever to a greater extent to _not_ advertise what's running _if they elect to do so_. There are still identifying leaks in the syndication scaffold (`generator`) that give the game away, and that's a core file so not so easy to modify sustainably (though https://github.com/textpattern/textpattern/issues/1761 may take care of that in due course).

Some folks like sharing what they run, some don't. It should be up to them to decide their level of involvement, depending on attack surface, paranoia level, and other factors.

**Comment by vanmelick** [comment]
Hey all,

On 2021-10-28 04:38, Philippe wrote:
>> /.well-know/change-password that would redirect to
>> textpattern/?reset=1
> 
> This is mostly for the benefit of password managers: to make it easier
> for the user to update/modify a password.

If it's for changing/updating the password, shouldn't the URL be:
/textpattern/index.php?event=admin&step=new_pass_form

Btw, if you don't want outsiders to know the TXP version, why display it 
in the RSS/Atom feed (generator field)
https://release-demo.textpattern.co/rss
https://release-demo.textpattern.co/atom

Not saying you should "fix" that, just that finding out which CMS is 
used is already very easy.

Regards,
--Ruud

**Comment by petecooper** [comment]
>Not saying you should "fix" that, just that finding out which CMS is used is already very easy.

That may be partly addressed or mitigated by https://github.com/textpattern/textpattern/issues/1761

There's a dead-cert way to find out a) the CMS and b) the version from a certain file, too - no auth required.

**Comment by phiw13** [comment]
> If it's for changing/updating the password, shouldn't the URL be:
/textpattern/index.php?event=admin&step=new_pass_form

That is / might be an option. Note that it gives away even more info about the internal workings of Textpattern. 

On the other hand, if the user has doubt about the password, it might be better to send them to the more external form I pointed to in my original idea (no need to login etc).

**Comment by vanmelick** [comment]
On 2021-12-27 02:09, Philippe wrote:
>> If it's for changing/updating the password, shouldn't the URL be:
>> /textpattern/index.php?event=admin&step=new_pass_form
> 
> That is / might be an option. Note that it gives away even more info
> about the internal workings of Textpattern.
> 
> On the other hand, if the user has doubt about the password, it might
> be better to send them to the more external form I pointed to in my
> original idea (no need to login etc).

The purpose of /.well-known/change-password is to help password 
management tools in finding the correct location to CHANGE a currently 
known password. It's not intended for finding a password RESET option, 
see:
https://w3c.github.io/webappsec-change-password-url/

Password management tools do not and cannot reset passwords. In most 
cases the reset procedure involves clicking a link in a 
password-reset-confirmation email to complete the reset. I have yet to 
see a password management tool that can complete such a reset.

Only humans reset passwords and they don't visit "/.well-known" urls, 
because it's far easier for them to simply navigate to the reset option 
themselves via the account login page.

--Ruud

**Comment by petecooper** [comment]
Tentatively closing for now as there's no clear consensus but we have a few potential routes in to address (e.g. web server rewrite rule). Please re-open if I've misunderstood sentiment. Thank you.
